### PR TITLE
[Dy2stat] Support that do not convert a function when use Dynamic-to-Stat

### PR DESCRIPTION
**Pull request by liym27**
### PR types

New features
### PR changes

Others
### Describe

- ### 功能
新增 `paddle.jit.not_to_static` 装饰器，可在动静转换场景中装饰函数，被装饰的函数将不会被转化。

- ### 背景需求：
1. 从动静转换功能的完整性和易用性层面考虑，需要支持 `not_to_static` 接口，以使得动转静过程中，特定的函数不被转化。
2. 尤其是当函数本身可以在静态图模型下使用，例如，函数内部本身就写了动态图和静态图2个分支，这类函数是无需再被转化成静态图函数的。[目前，这种情况在 PaddleDetection 模型中尤为常见]

- ### 使用场景：
前提：一个函数本就可以在静态图模式下成功运行，在以下场景中可以使用 `not_to_static` 装饰该函数。
1.  该函数在转化过程中出现未知错误；
2. 该函数内部实现中，本身具有动态图分支和静态图2种分支；
3. 用户不想在动转静过程中转化该函数；
--- 
Add a decorator  `paddle.jit.not_to_static` to support that do not convert a function in Dynamic-to-Static.

- ### Background
1. For the integrity of Dynamic-to-Static function, add a decorator to suppresses the convertion of a spectial function. 
2. Especially when the function itself can be used in static mode, for example, there are two branches of static mode and static mode in the function.

- ### Usage scenarios：
1. An unknown error occurs in the dynamic-to-static conversion process of the function;
2. In the internal implementation of the function, it has two branches: dynamic branch and static branch;
3. Users don't want to convert the function in the process of dynamic to static;

For example, when you don't want to convert a function when use Dynamic-to-Static, you can use `@paddle.jit.not_to_static` to decorate the function. 

Here, `func_not_to_static` will not be converted. Note: `func_not_to_static` can be executed in dynamic mode or static mode.
```python
import paddle

@paddle.jit.not_to_static
def func_not_to_static(x):
    res = x - 1
    return res

@paddle.jit.to_static
def func(x):
    if paddle.mean(x) < 0:
        x_v = func_not_to_static(x)
    else:
        x_v = x + 1
    return x_v

x = paddle.ones([1, 2], dtype='float32')
x_v = func(x)
print(x_v) # [[2. 2.]]
```

---
Document of `not_to_static`
http://gzhxy-inf-bce36a39de.gzhxy.baidu.com:8121/documentation/docs/zh/api/index_cn.html?reviewVersion=jenkins-doc_review-180
image

---

- [ ] TODO: Update document of **[Dygraph to Static Graph](https://www.paddlepaddle.org.cn/documentation/docs/en/develop/guides/04_dygraph_to_static/basic_usage_en.html#programtranslator)**

### CODE REVIEW

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py [hunk 1]
```diff
@@ -35,13 +33,57 @@
...
     collections, pdb, copy, inspect, re, six, numpy, logging
 ]
 
 translator_logger = TranslatorLogger()
 
+DISABLE_CONVERTION = "An attribute for a function that indicates that" \
+                  "the user does not want it to be converted in dynamic-to-static."
+
+
+def not_to_static(func=None):
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: The name `not_to_static` may be too long. Could we consider `ignore` or `skip` or other single word as decorator name?

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py [hunk 2]
```diff
@@ -133,6 +175,13 @@ def dyfunc(x):
     # in this case, unwraps it into a raw method or function.
     _, func = unwrap_decorators(func)
 
+    if getattr(func, DISABLE_CONVERTION, False):
+        translator_logger.log(
+            2,
+            "{} is not converted it is decorated by 'paddle.jit.not_to_static'.".
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: `"{} is not converted when it is decorated by 'paddle.jit.not_to_static'."` add `when`

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py [hunk 3]
```diff
@@ -175,14 +180,104 @@ def test_transformed_static_result(self):
...
+    return y
+
+
+class TestClass(paddle.nn.Layer):
+    @paddle.jit.not_to_static
+    def called_member(self, x):
+        return paddle.sum(x)
+
+    @declarative
+    def forward(self, x):
```
Comment (Aurelius84) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py:212]: Not important, use `to_static` to replace `declarative`. In the future, we will deprecate all usage of `declarative`.

**Review by Aurelius84 — COMMENTED** [review]
Great work!

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 2 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: Thanks, done

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py (hunk 3 above)
Comment (liym27) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py:212]: Thanks, done

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 1 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: Thanks for your suggestion!

`ignore` or `skip` can't express the complete semantic well, which may make users confused, and `not_to_static` can express the meaning more clearly and accurately. 
If `not_to_static` is too long, maybe we could discuss about it in more detail.

Comment (zhhsplendid) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: Here I agree with Yamei. Ignore or skip doesn't express the full meaning

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py [hunk 4]
```diff
@@ -35,13 +33,71 @@
...
+        not_convert(bool): An attribute indicates that the function won't be converted in dynamic-to-static.
+
+    NOTE(liym27): More attributes and methods can be added in this class.
+    """
+
+    def __init__(self, not_convert=False):
+        self.not_convert = not_convert
+
+
+def not_to_static(func=None):
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: Should we put `not_to_static` at same place of `to_static`? I think that's more normal.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py [hunk 5]
```diff
@@ -175,14 +179,106 @@ def test_transformed_static_result(self):
...
+        y = self.called_member(x)
+        return y
+
+
+class TestNotToConvert(TestRecursiveCall2):
+    def set_func(self):
+        self.dygraph_func = func_not_to_static
+
+    def test_CONVERSION_OPTIONS(self):
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py]: Naming should be `test_conversion_options`

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 4 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py]: Yes, done. Thanks.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py (hunk 5 above)
Comment (liym27) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_convert_call.py]: Thanks, done.

**Review by zhwesky2010 — APPROVED** [review]
lgtm for unittest change name

**Review by Aurelius84 — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]
LGTM

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**